### ISSUE TITLE
What is the order in which the `auras` are loaded? + Ideas to make class usage possible

### ISSUE THREAD

**Issue by Vrakfall** [WeakAuras/WeakAuras2#809]
Hello,

I'd like to know in which order the `auras` are loaded. I'll explain why.

I'm currently trying to re-use huge chunks of my code I made by using closure-based (for faster access) classes (actually objects, the closest thing to classes `LUA` can offer us). For that I thought of multiple ways to implement it and I'd like more information in order to see if my main idea can work before resolving to using a fallback idea.

To achieve this main idea, I'm loading my classes (actually functions returning an object based on a prototype and the arguments it receives) in global variables in the `On Init` `action` of an `aura` that is always loaded and has its only trigger set to `Status -> Conditions -> Always active trigger`. I then instanciate the variables with the classes used in each `aura` within their own `On Init` `action`. I know, this is a bit ugly and it breaks to portability of the `auras` made that way but the goal was really to re-use the code: Writing the behaviour of some animations acting differently depending on the arguments set at the instanciation, all of that in order to avoid the need to copy-paste the used classes in the `On Init` of each `aura`. Doing that, on top of being a dull copy-paste work, would mean a class "definition" is actually called multiple times (and overwriting the variable each time, which is useless).

This works fine when using it in another `aura` for a debuff spell (like `Agony`) with a trigger of the `aura` type: the class is accessible, instanciated and used without problem. That meant for me the `auras` were loaded in the right order. I thought for a moment that the ordering was based on the name, as appearing in the option panel, when many `auras` would be loaded during the same frame.

But then I tried to re-use a class in an `aura` for a normal spell with a long cooldown (like `Phantom Singularity`) by using a `status` type trigger (a `Cooldown Progress (Spell)` one) and it didn't work. After further debugging, it was pretty obvious the `auras` weren't loaded in the order I expected, as this one was always loaded before the one with the classes and the `status` type trigger. I tried changing the names, changing the trigger of the class-loading `aura` (but in a way it always loads at startup, which I only tried with `status` ones),... Nothing I tried made the ordering change even though both have a `status` type trigger.

That made me wonder if the type of trigger is important for the loading order of the `auras`, is it? Also, is it dependent of the events and the order they are fired up (like when logging in `WoW`)? Does the trigger type actually matter for the load ordering when loading conditions are actually completely different?

I suppose I could potentially "define" my classes in the `On Init` section of a group which should potentially always be loaded before the `auras` it contains but, as groups can't contain other groups, it wouldn't allow me to make different groups for, i.g., different classes of characters (warlock, priest,...) or other sorting needs.

I thought of a feature that could be useful for this: An `aura` that could be defined as a `dependency` for other `auras`, a bit like how libraries are used in coding. The `auras` could tell they need the dependency ones as such, which could, as an implementation example, include the dependency in their export string. Then, if multiple `auras` have the same dependency, that one is actually imported only once in order to avoid redundancy.

Another implementation possibility for that last idea would be a special type of `aura`, like a `class` type, which would be loaded before any other type of `aura`, allowing for a unique location for a user to store its class definitions.

I know, this looks like all about re-using patterns already used in programmation but this could actually really help users like me to keep a more maintainable code throughout multiple and numerous `auras`. This is still easier and way faster than making a full add-on.

Thank you really much for reading it all and don't hesitate to criticise.

**Comment by emptyrivers** [comment]
Auras are loaded via an iteration through the database with `pairs`. In particular, this means that there is no defined order.

I think that if you are at the point that the load order of your displays (which are kind of supposed to be independent of each other) is a problem, then perhaps you should consider offloading the bulk of your code to a personal library addon, and using WeakAuras as the visible portion of your setup.

**Comment by Vrakfall** [comment]
That's a possibility. But then, if I were to share the displays with other people, I would have to ask them to download my addon. That's another story than just giving a few export strings. That would also require potential maintenance throughout expansions.

**Comment by emptyrivers** [comment]
It's what I do with https://wago.io/NormalStagger and https://wow.curseforge.com/projects/brewmastertools. Works pretty well.

**Comment by Vrakfall** [comment]
Well, thank you. I'll start looking into that.

Still, it would be nice to be able to share things between displays, or be able to make sub-groups for better sorting. Sub-groups could actually allow for better sharing between big groups of displays.

Being able to share code more easily would also make it less necessary to dig up in the whole process of making an addon.
